mmgen/models/architectures/sngan_proj/modules.py

```diff
@@ -0,0 +1,236 @@
+                                       **shortcut_cfg)
+
+    def forward(self, x):
+        out = self.activate(x)
+        out = self.conv_blocks[0](out)
+        out = self.activate(out)
+        out = self.conv_blocks[1](out)
+        if self.downsample:
+            out = self.downsample(out)
+
+        shortcut = self.forward_shortcut(x)
+        return out + shortcut
+
+    def forward_shortcut(self, x):
+        out = x
+        if self.learnable_sc:
+            out = self.shortcut(out)
+            if self.downsample:
+                out = self.downsample(out)
+        return out
+
+    def _weight_init(self):
+        xavier_init(self.conv, gain=np.sqrt(2), distribution='uniform')
+        if self.learnable_sc:
+            xavier_init(self.shortcut, gain=1, distribution='uniform')
+
+
+@MODULES.register_module()
+class SNGAN_DisHeadResBlock(nn.Module):
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 conv_cfg=None,
+                 shortcut_cfg=None,
+                 act_cfg=dict(type='ReLU')):
+
+        super().__init__()
+        self.activate = build_activation_layer(act_cfg)
+        conv_blocks = [
+            ConvModule(in_channels, out_channels, **conv_cfg),
+            ConvModule(out_channels, out_channels, **conv_cfg)
+        ]
+        self.conv_blocks = nn.ModuleList(conv_blocks)
+
+        self.downsample = nn.AvgPool2d(2, 2)
+
+        self.shortcut = ConvModule(in_channels, out_channels, **shortcut_cfg)
+
+    def forward(self, x):
+        out = self.conv_blocks[0](x)
+        out = self.activate(out)
+        out = self.conv_blocks[1](out)
+        out = self.downsample(out)
+
+        shortcut = self.forward_shortcut(x)
+        return out + shortcut
+
+    def forward_shortcut(self, x):
+        out = self.downsample(x)
+        out = self.shortcut(out)
+        return out
+
+    def _weight_init(self):
+        xavier_init(self.conv, gain=np.sqrt(2), distribution='uniform')
+        if self.learnable_sc:
+            xavier_init(self.shortcut, gain=1, distribution='uniform')
+
+
+@MODULES.register_module()
+class SNcBatchNorm(nn.Module):
+
+    def __init__(self,
+                 in_channels,
+                 num_classes,
+                 use_cbn=True,
+                 norm_cfg=dict(type='BN'),
+                 use_norm_affine=False,
+                 auto_sync_bn=True):
+        super().__init__()
+        norm_cfg = deepcopy(norm_cfg)
+        norm_type = norm_cfg['type']
+
+        if norm_type not in ['IN', 'BN', 'syncBN']:
+            raise ValueError('Only support `IN` (InstanceNorm), '
+                             '`BN` (BatcnNorm) and `syncBN` for '
+                             'Class-conditional bn. '
+                             'Receive norm_type: {}'.format(norm_type))
+        norm_cfg['affine'] = use_norm_affine
+        _, self.norm = build_norm_layer(norm_cfg, in_channels)
+
+        if auto_sync_bn and check_dist_init():
```

Comment:
I may recommend that you can build SyncBN with build_norm_layer. 

```python
if auto_sync_bn and check_dist_init() an d norm_cfg['type'] == `BN`:
    _norm_cfg = deepcopy(norm_cfg)
    _norm_cfg['type'] = 'SyncBN'
   # build norm layer with _norm_cfg

```

Reply:
Maybe we do not need `deepcopy` here?
We have already `deepcopy` the input config at the start of the function.